### PR TITLE
Add RSpec/AnyInstance

### PR DESCRIPTION
**Pull request by mlarraz**
Because `allow_any_instance_of` and `expect_any_instance_of` are bad

### CODE REVIEW

**Comment by andyw8** [comment]
This would be a nice addition! :+1:

File: spec/project_spec.rb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # encoding: utf-8
 
-describe 'RuboCop Project' do # rubocop:disable RSpec/DescribeClass
+describe 'RuboCop Project' do
```
Comment (andyw8) [inline spec/project_spec.rb]: This change seems be causing Travis to fail.

Comment (mlarraz) [inline spec/project_spec.rb]: Yeah interestingly I actually need that to make it pass locally, due to
different versions of Rubocop I guess.

**Comment by mlarraz** [comment]
I believe this is ready to go now

**Comment by pd** [comment]
:+1: I would love to see this make it into master, I am just itching to properly ban `any_instance` from the projects I work on without having to scan for it in every pull request.

**Comment by packrat386** [comment]
:+1: This would be great to have. I've seen too many bad tests with it.

**Comment by nijikon** [comment]
@mlarraz Please rebase on top of master

**Comment by nijikon** [comment]
@mlarraz Can you please rebase again, add changelog entry and squash to one commit? Sorry for the inconvenience.

**Comment by nijikon** [comment]
Still missing the changelog entry and commits need to be squashed to one.

**Comment by nijikon** [comment]
Nearly there, please fix tests :)

**Comment by mlarraz** [comment]
Thanks, don't think I've ever seen a project with tests for their changelog before

**Comment by nijikon** [comment]
@mlarraz This is taken from rubocop :) https://github.com/bbatsov/rubocop/blob/master/spec/project_spec.rb

**Comment by maciejmajewski** [comment]
Could you explain why it is a bad practice?

**Comment by andyw8** [comment]
https://relishapp.com/rspec/rspec-mocks/docs/working-with-legacy-code/any-instance

> This feature is sometimes useful when working with legacy code, though in general we
> discourage its use for a number of reasons:
> * The rspec-mocks API is designed for individual object instances, but this feature operates on entire classes of objects. As a result there are some semantically confusing edge cases. For example, in expect_any_instance_of(Widget).to receive(:name).twice it isn't clear whether a specific instance is expected to receive name twice, or if two receives total are expected. (It's the former.)
> *Using this feature is often a design smell. It may be that your test is trying to do too much or that the object under test is too complex.
> *It is the most complicated feature of rspec-mocks, and has historically received the most bug reports. (None of the core team actively use it, which doesn't help.)

**Comment by maciejmajewski** [comment]
@andyw8 thanks

**Comment by nijikon** [comment]
:+1:

**Comment by dpisarewski** [comment]
Could you explain why stubbing of the `new` method is better? In my opinion it's much worse than stubbing a needed method of the class.

**Comment by bquorning** [comment]
> In my opinion it's much worse than stubbing a needed method of the class.

What `any_instance` does is not stubbing a needed method of the class. It’s stubbing a needed method of *an instance of* the class. I like that by stubbing `new` you now at least know which object instance you are working with.

But, as mentioned above:

> Using this feature is often a design smell. It may be that your test is trying to do too much or that the object under test is too complex.

Stubbing `new` obviously doesn’t fix this problem.

**Comment by backus** [comment]
Here is an example of why people usually say that `any_instance` should be the very last tool you reach for to test your code:

```ruby
require 'rspec'

class Foo
  def bar
    :original_implementation
  end
end

$global_instance = Foo.new

RSpec.describe Foo do
  context 'any instance of' do
    before do
      allow_any_instance_of(Foo)
        .to receive(:bar)
        .and_return(:stubbed_implementation)
    end

    it 'stubs method for instance created in life cycle' do
      expect(Foo.new.bar).to be(:stubbed_implementation)
    end

    it 'hijacks instances created without Foo.new' do
      expect(Class.instance_method(:new).bind(Foo).call.bar)
        .to be(:stubbed_implementation)
    end

    it 'stubs preexisting instance created outside of test' do
      expect($global_instance.bar).to be(:stubbed_implementation)
    end
  end

  context 'stubbing .new' do
    before do
      stubbed_instance = Foo.new

      allow(stubbed_instance)
        .to receive(:bar)
        .and_return(:stubbed_implementation)

      allow(Foo)
        .to receive(:new)
        .and_return(stubbed_instance)
    end

    it 'stubs method for instance created in life cycle' do
      expect(Foo.new.bar).to be(:stubbed_implementation)
    end

    it 'does not hijack objects created without Foo.new' do
      expect(Class.instance_method(:new).bind(Foo).call.bar)
        .to be(:original_implementation)
    end

    it 'does not stub preexisting instance created outside of test' do
      expect($global_instance.bar).to be(:original_implementation)
    end
  end
end
```

```
$ rspec --format documentation any_instance_madness.rb

Foo
  any instance of
    stubs method for instance created in life cycle
    hijacks instances created without Foo.new
    stubs preexisting instance created outside of test
  stubbing .new
    stubs method for instance created in life cycle
    does not hijack objects created without Foo.new
    does not stub preexisting instance created outside of test

Finished in 0.00695 seconds (files took 0.07858 seconds to load)
6 examples, 0 failures
```

Summarizing what this code example shows:

1. Stubbing `.new` ensures that the object you are stubbing is created during the life cycle of your test.
2. Stubbing `.new` specifies that your object is created through `.new`.

**Comment by dpisarewski** [comment]
Thank you for the explanation.

**Pull request merged**